src/conn/conn_sweep.c

```diff
@@ -129,32 +111,110 @@ __sweep(WT_SESSION_IMPL *session)
 		    !__wt_txn_visible_all(session, btree->rec_max_txn))
 			goto unlock;
 
-		/* If the handle is open, try to close it. */
-		if (F_ISSET(dhandle, WT_DHANDLE_OPEN)) {
+		if (F_ISSET(dhandle, WT_DHANDLE_OPEN))
 			WT_WITH_DHANDLE(session, dhandle, ret =
-			    __wt_conn_btree_sync_and_close(session, 0, 0));
-
-			/* We closed the btree handle, bump the statistic. */
-			if (ret == 0)
-				WT_STAT_FAST_CONN_INCR(
-				    session, dh_conn_handles);
-		}
-
-		if (dhandle->session_inuse == 0 && dhandle->session_ref == 0)
-			++closed_handles;
+			    __wt_conn_btree_sync_and_close(session, 0, 1));
 
 unlock:		WT_TRET(__wt_writeunlock(session, dhandle->rwlock));
 		WT_RET_BUSY_OK(ret);
 	}
 
-	if (closed_handles) {
-		WT_WITH_DHANDLE_LOCK(session,
-		    ret = __sweep_remove_handles(session));
+	return (0);
+}
+/*
+ * __sweep_flush --
+ *	Flush pages from dead trees.
+ */
+static int
+__sweep_flush(WT_SESSION_IMPL *session)
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_DATA_HANDLE *dhandle;
+	WT_DECL_RET;
+	time_t now;
+
+	conn = S2C(session);
+
+	/* Don't discard handles that have been open recently. */
+	WT_RET(__wt_seconds(session, &now));
+
+	WT_STAT_FAST_CONN_INCR(session, dh_conn_sweeps);
+	SLIST_FOREACH(dhandle, &conn->dhlh, l) {
+		if (!F_ISSET(dhandle, WT_DHANDLE_DEAD))
+			continue;
+
+		/*
+		 * We have a candidate for closing; if it's open, acquire an
+		 * exclusive lock on the handle and close it.
+		 */
+		if ((ret =
+		    __wt_try_writelock(session, dhandle->rwlock)) == EBUSY)
+			continue;
 		WT_RET(ret);
+
+		/* If the handle is marked "dead", flush it from cache. */
+		WT_WITH_DHANDLE(session, dhandle, ret =
+		    __wt_conn_btree_sync_and_close(session, 0, 0));
+
+		/* We closed the btree handle, bump the statistic. */
+		if (ret == 0)
+			WT_STAT_FAST_CONN_INCR(session, dh_conn_handles);
+
+		WT_TRET(__wt_writeunlock(session, dhandle->rwlock));
```

Comment:
Can't we be here when `WT_DHANDLE_OPEN` is not set?

In which case, `__wt_conn_btree_sync_and_close` isn't going to do anything.